DOCS-11576: LLM Obs span/trace query page (#30725)
DataDog/documentation: content/en/llm_observability/monitoring/querying.md

---
title: Querying spans and traces
further_reading:
  - link: "tracing/trace_explorer/query_syntax/"
    tag: "Documentation"
    text: "Trace Explorer Query Syntax"
---

## Overview
This page discusses using Datadog's [LLM Observability Trace Explorer][1] to query your LLM application's spans and traces.

> In Datadog, a _span_ represents an operation, and a _trace_ represents a request. Traces consist of one or more nested spans. For more information about this terminology, see [LLM Observability Terms and Concepts][2].

Reviewer: ```suggestion
In Datadog, a _span_ represents a unit of work representing a single operation in your LLM application. A _trace_ represents the end-to-end operations involved in processing a request in your LLM application, often consisting of one or more nested spans. For more information about this terminology, see [LLM Observability Terms and Concepts][2].
```

> To toggle between searching spans and traces, use the drop-down menu on the left of the query field. When searching traces, a query returns traces where the root span matches the query.

Reviewer: ```suggestion
To toggle between searching spans and traces, use the drop-down menu on the left of the query field on [LLM Observability Trace Explorer][1]. To search across traces where the root span matches the query, select `Traces`. To search across all your spans, including nested child spans, select `All Spans`. 
```

### Query by attribute

> _Span attributes_ are information collected from a span itself. Attribute queries take the form `@key:value`. All attribute keys are prepended with `@`.

Reviewer: ```suggestion
_Span attributes_ are structured, key-value metadata attached directly to each span, capturing details about the span’s execution, such as performance metrics, resource identifiers, or parameter values. 

Attribute queries take the form `@key:value`. All attribute keys are prepended with `@`.
```

> For example, the query `@duration:>5s` returns spans that took more than 5 seconds to complete.

Author: My goal here was to put example queries high enough on the page that a fly-by reader can immediately see one and understand the general structure of a query. I'd like to keep these inline examples and just repeat them in the example section.

### Query by tag

> _Span tags_ are information collected from the context around a span. Tag queries take the form `key:value`. In contrast to attribute keys, tag keys are not prepended with `@`.

Reviewer: ```suggestion
_Span tags_ are high-level key-value pairs used to group, segment, and correlate telemetry data across spans, services, or environments. Tags often indicate broader context, such as application name, environment, or deployment region, and are attached to spans to enable efficient search and aggregation.

Tag queries take the form `key:value`. In contrast to attribute keys, tag keys are not prepended with `@`.
```

For example, the query `ml_app:my_llm_app` returns spans from an application named `my_llm_app`.

### Query LLM input and output

> You can also search for free text in the input and output of your LLM application's spans. Wrap your free text query with `"`.

Reviewer: ```suggestion
You can also search for specific keywords, phrases, or strings on any span that has an input and output pair. To use free text search, wrap your free text query with `"`.
```

> For example, the query `"what's the weather"` returns spans that contain the string `what's the weather` in the input or output.

Reviewer: ```suggestion
For example, the query `"what's the weather"` returns agent, workflow, or LLM spans that contain the string `what's the weather` in the input or output.
```

<div class="alert alert-info">Free text queries are limited to the first 20,500 characters of a trace's input or output.</div>

### Operators

You can combine multiple search terms using the boolean operators `AND` (intersection), `OR` (union), and `-` (exclusion).

For example, the query `@duration:>5s AND -"foo"` returns spans that took more than 5 seconds to complete and do **not** contain the string `foo` in the input or output.

### Query syntax

The LLM Observability Trace Explorer shares the same query syntax as Datadog's [APM Trace Explorer][6]. For query syntax details, including wildcard search, handling numerical values, escaping special characters, and more, see [Trace Explorer Query Syntax][6].

## Example queries

`@status:error`

> : Finds spans or traces that have a status of `error`.

Reviewer: can we lead with "find" rather than "finds"

Reviewer: for each of these, a table with query and match description like this would be better: https://docs-staging.datadoghq.com/cswatt/llm-tracing/tracing/trace_explorer/query_syntax/#tags-search

### Evaluation queries

Use the `@evaluations` attribute to find spans or traces by [evaluation][3] result.

`@evaluations.failure_to_answer:"failed to answer"`
: Finds spans or traces where the LLM failed to deliver an appropriate response.

`@evaluations.hallucination:"hallucination found"`
: Finds spans or traces where a hallucination is detected.

`@evaluations.input_sentiment:negative`
: Finds spans or traces with negative input sentiment.

`@evaluations.input_toxicity:toxic`
: Finds spans or traces where harmful or inappropriate content is detected in the input.

`@evaluations.language_mismatch:mismatched`
: Finds spans or traces where the user's prompt and the LLM's response are in different languages.

`@evaluations.output_sentiment:negative`
: Finds spans or traces with negative output sentiment.

`@evaluations.output_toxicity:toxic`
: Finds spans or traces where harmful or inappropriate content is detected in the output.

`@evaluations.prompt_injection:found`
: Finds spans or traces where prompt injection is detected.

`@evaluations.topic_relevancy:irrelevant`
: Finds spans or traces where the prompt input diverges from the LLM application's intended topic.

> You can also search spans by the results of [custom evaluations][4]. For example, if you have a custom evaluation called `user_mood` with categorical values `happy`, `sad`, and `tired`, you could use the query: `@evaluations.custom.user_mood:happy`.

Reviewer: can we move this to the top of this section or have it be a tab between Custom Evaluations and Datadog Evaluations?

### Metadata queries

Use the `@meta` attribute to find spans by metadata information.

`@meta.span.kind:<span_kind>`
: Finds spans by [_span kind_][5]. For example, `<span_kind>` could be `llm`, `workflow`, `tool`, etc. 

`@meta.model_provider:openai`
: Finds spans or traces where the model provider is OpenAI.

`@meta.model_name:gpt-4.1`
: Finds spans or traces where the model is GPT-4.1.

### Trace-level queries

> To search traces based on attributes of the spans that these traces contain, use the `@child` attribute.

Reviewer: ```suggestion
To search traces based on attributes of its nested spans, use the `@child` attribute.
```
